### PR TITLE
apache: bump version plus brush up

### LINKED ISSUE #3720: Apache: do not list modules
I installed Apache and configured after reading this https://wiki.openwrt.org/doc/howto/http.apache
Is running fine.

httpd -M or apachectl -M in the shell should list installed modules. All i get is "Syntax OK"

The response should be something like this:
Loaded Modules:
core_module (static)
mpm_prefork_module (static)
http_module (static)
so_module (static)
auth_basic_module (shared)
cgi_module (shared)
version_module (shared)
wsgi_module (shared)
Syntax OK

http://www.shellhacks.com/en/List-Loaded-Enabled-Apache-Modules

### LINKED ISSUE #4840: apache proxy segfault
Maintainer @heil 
with --enable-proxy on sun7i target (arm) I get this error and it will loop forewer

```
[Sat Sep 16 13:52:16.047623 2017] [proxy_hcheck:info] [pid 34359738384:tid 3064544580] (70023)This function has not been implemented on this platform: AH03312: apr_thread_pool_create() with 16 threads failed
[Sat Sep 16 13:52:16.047849 2017] [watchdog:debug] [pid -5264627648507674620:tid 3069363352] mod_watchdog.c(585): AH02981: Watchdog: Created worker thread (_proxy_hcheck_).
[Sat Sep 16 13:52:16.048519 2017] [proxy:debug] [pid -5264506015033851904:tid 3069363352] proxy_util.c(1777): AH00925: initializing worker proxy:reverse shared
[Sat Sep 16 13:52:16.048586 2017] [proxy:debug] [pid -5264506015033851904:tid 3069363352] proxy_util.c(1819): AH00927: initializing worker proxy:reverse local
[Sat Sep 16 13:52:17.043593 2017] [core:notice] [pid -4715750481822642040:tid 3069363352] AH00052: child pid 10645 exit signal Segmentation fault (11)
[Sat Sep 16 13:52:17.044001 2017] [core:notice] [pid -4715750481822642040:tid 3069363352] AH00052: child pid 10646 exit signal Segmentation fault (11)
[Sat Sep 16 13:52:17.044068 2017] [core:notice] [pid -4715750481822642040:tid 3069363352] AH00052: child pid 10647 exit signal Segmentation fault (11)
[Sat Sep 16 13:52:17.044124 2017] [core:notice] [pid -4715750481822642040:tid 3069363352] AH00052: child pid 10648 exit signal Segmentation fault (11)
```

### PR DESCRIPTION
**Pull request by micmac1**
Maintainer: @heil 
Compile tested: master ath79, 19.07-rc2 ath79+arc770
Run tested: 19.07-rc2 ath90, it starts fine and I get to an error page when visiting the server with my browser

Description:
Hi all,

I guess I got a bit bored over the holidays, honestly.

I'm hoping that this

Fixes #4840 
Fixes #3720

Well, the modules are now installed. I also incorporated a layout file like @minfrin talked about in #10883 (I used his file in the beginning but wasn't totally satisfied, so I made some amendments).

Regarding the segfault when using proxy mode. Well, I don't know much about web servers. So I'm hoping that @lucize, @0anton, @tru7, @bilicio or @jpereira can confirm whether this is fixed.

I also added a simple init script.

Kind regards,
Seb

### CODE REVIEW

**Comment by minfrin** [comment]
Confirm the changes to layout for me? I’m keen to get openwrt layout formally into httpd, next step is for me to propose it for backport.

Making similar changes to apr-util to give us all the modules.

**Comment by micmac1** [comment]
> Confirm the changes to layout for me? I’m keen to get openwrt layout formally into httpd, next step is for me to propose it for backport.

Look at 726b61f for the layout I used.

To me it seems that using a layout brings you a bit closer to where you want to be, but it doesn't take you all the way. For instance you can define  proxycachedir in the layout, but this variable is not evaluated in the layout setup, which means you can't append '+' (well, you can, but it won't be replaced with the target).

And then there are other variables that do get properly evaluated, like sysconfdir, but whatever comes out of this gets overwritten by configure args. And '--sysconfdir=/etc' is in OpenWrt's CONFIGURE_ARGS by default, so then the layout output for this var is overwritten. You end up having to specify it in CONFIGURE_ARGS manually.

I'm not sure that this needs to be upstreamed to apache. It looks like Alpine for instance also didn't upstream their layout but instead just keep it locally.

**Comment by micmac1** [comment]
Well, I thought the error I saw in my web browser was normal. It doomed on me a bit later that really I should be seeing the infamous "It works" page.

I ended up removing two patches and adding a substitute which I "borrowed" from our friends at buildroot (thanks!). Now when I install just the base package and enter the router's LAN IP into my browser I see that "It works" :) So that's a success I guess.

I added some other small amendments to the last commit as well. I can squash the commits later. Who knows, maybe this won't be the last edit ;)

**Comment by neheb** [comment]
https://github.com/openwrt/packages/pull/10031

Can you grab the OpenSSL patch from there?

Why is a binary being installed in InstallDev?

**Comment by micmac1** [comment]
> #10031
> 
> Can you grab the OpenSSL patch from there?

I won't sign-off on any OpenSSL patches. The reason is that I'm not a programmer. I can't vouch for stuff like that and when it comes to crypto this is too sensitive IMHO to take lightly. I suggest to get these changes integrated upstream first before applying them to OpenWrt.

> Why is a binary being installed in InstallDev?

It's not really a binary. It's a script and apparently needed when you compile out-of-tree modules. Both buildroot and Alpine package it in their dev packages.

Kind regards,
Seb

**Comment by minfrin** [comment]
> 726b61f

Can you verify that this is ok - http://svn.apache.org/viewvc?rev=1872093&view=rev - I will backport this if correct.

I'm keen that Apache in the distros is working as pain free as possible, I would like to reach a point where there are (almost) no patches for httpd in the patches directory, and the deployment have full module support.

Can you explain in more detail whether there is anything that needs to be fixed upstream with respect to layout? I'm not seeing that --sysconfdir=/etc being overriden is a problem, we can live with that, what is a problem is if the  values contradict each other.

**Comment by minfrin** [comment]
> I won't sign-off on any OpenSSL patches. The reason is that I'm not a
> programmer. I can't vouch for stuff like that and when it comes to crypto
> this is too sensitive IMHO to take lightly. I suggest to get these changes
> integrated upstream first before applying them to OpenWrt.

The following patch to trunk has been proposed for backport to v2.4.x, when it's accepted this problem will go away:

http://svn.apache.org/viewvc?view=revision&revision=1861950

If you're willing to wait for 2.4.42 (which is sensible) this problem should go away.

**Comment by minfrin** [comment]
Actually the openssl patch is doing something different, and hasn't been applied to httpd trunk (patches of patches make my eyes bleed) - +1 to whoever having contributed the patch contribute it upstream, I will sort it out. @neheb

**Comment by neheb** [comment]
@minfrin try compiling with -DOPENSSL_API_VERSION=0x10100000L and -Werror=implicit-function-declaration. The patch is not perfect but fixes such a situation.

**Comment by minfrin** [comment]
Please submit the patch to issues.apache.org, or post it to the dev@httpd.apache.org.

The patch is not my intellectual property, so I can’t commit it to httpd without you having given your say so first, and that happens when you submit it above.

As soon as you do that I can get it committed upstream for you, then get it back ported.

**Comment by micmac1** [comment]
> > [726b61f](https://github.com/openwrt/packages/commit/726b61fb6e92b870c9c1712b3169fe199f9447d5)
> 
> Can you verify that this is ok - http://svn.apache.org/viewvc?rev=1872093&view=rev - I will backport this if correct.

This looks OK to _me_ ;) Let's wait a bit to see what gets checked in here in the end. Small nitpick: It's OpenWrt, not OpenWRT (I have to fix the same in my local layout).

> I'm keen that Apache in the distros is working as pain free as possible, I would like to reach a point where there are (almost) no patches for httpd in the patches directory, and the deployment have full module support.

That is certainly commendable :)

> Can you explain in more detail whether there is anything that needs to be fixed upstream with respect to layout? I'm not seeing that --sysconfdir=/etc being overriden is a problem, we can live with that, what is a problem is if the values contradict each other.

I'm not sure I would call it a contradiction. I just find it a bit annoying that proxycachedir isn't treated the same way as - say - runtimedir with regards to '+'. See build/apr_common.m4. It lists the variables that get this special treatment.

```
  for var in prefix exec_prefix bindir sbindir libexecdir mandir \
             sysconfdir datadir includedir localstatedir runtimedir \
             logfiledir libdir installbuilddir libsuffix $3; do
```

**Comment by micmac1** [comment]
> > Can you explain in more detail whether there is anything that needs to be fixed upstream with respect to layout? I'm not seeing that --sysconfdir=/etc being overriden is a problem, we can live with that, what is a problem is if the values contradict each other.
> 
> I'm not sure I would call it a contradiction. I just find it a bit annoying that proxycachedir isn't treated the same way as - say - runtimedir with regards to '+'. See build/apr_common.m4. It lists the variables that get this special treatment.
> 
> ```
>   for var in prefix exec_prefix bindir sbindir libexecdir mandir \
>              sysconfdir datadir includedir localstatedir runtimedir \
>              logfiledir libdir installbuilddir libsuffix $3; do
> ```

Mmh, just to make it clear, this is not an important issue for me :) I don't think anybody needs to make a fuzz about this.

**Comment by minfrin** [comment]
I fixed the case for you in http://svn.apache.org/viewvc?rev=1872102&view=rev.

+1 to fixing the special treatment. I've been working on httpd for decades and this is the first I've seen that it was a thing (!!).

**Comment by minfrin** [comment]
I will propose http://svn.apache.org/viewvc?rev=1872103&view=rev for backport if there are no objections upstream.

**Comment by minfrin** [comment]
This PR has been included in https://github.com/openwrt/packages/pull/10924, which makes it possible to build third party Apache modules.

**Comment by tru7** [comment]
Could successfully build everything ( and also test SSL and proxy module) for ath79 / mips_exect for mod_md. This fails like this:

checking for arc4random_buf... no
checking whether to enable mod_md... configure: error: mod_md has been requested but can not be built due to prerequisite failures

**Comment by micmac1** [comment]
> Could successfully build everything ( and also test SSL and proxy
> module) for ath79 / mips_exect for mod_md. This fails like this:
>
> checking for arc4random_buf... no
> checking whether to enable mod_md... configure: error: mod_md has been requested but can not be built due to prerequisite failures

mod_md depends on libcurl, jansson & libopenssl. Can you check in
menuconfig if they're all available?

**Comment by tru7** [comment]
thank you, that helped. I hadn't installed the feed for jansson. Now all selectable packages for apr-util and apache are successfully compiled.
I can't wait to get that PR merged.

**Comment by micmac1** [comment]
Added another sed script to fix apxs, so that it can be used to cross-compile external modules (we don't have any in the tree yet, but I guess there's interest for it). I setup a quick test and added an external package called mod_gnutls. It cross-compiled fine.

I'll squash the commits, probably tomorrow.

**Comment by minfrin** [comment]
For the record, assuming another two +1s, the apxs fix should be in httpd 2.4.42:

https://github.com/apache/httpd/blob/142e11fd6bb8f0cf6c7d9b7388dbea7cbe332c86/STATUS#L286

**Comment by micmac1** [comment]
I squashed the 9 commits into 6. I don't want to go for less than that because then it'd be a bit hard to follow.

I ended up removing the BROKEN from the two packages that depend on #10906. So please consider merging both (first the other pull request, then this one).

Kind regards,
Seb

**Comment by micmac1** [comment]
Hi all,

I checked the logs and it looks like maintainer @heil is inactive. I don't mind taking over maintainership for apr/apr-util/apache. If you're OK with that I'll add a corresponding commit.

@hnyman @neheb 

Kind regards,
Seb

**Comment by hnyman** [comment]
Just do it.  
(or alternatively, add yourself as the second maintainer)

**Comment by micmac1** [comment]
Added myself as a second. That's probably best. Thanks!

**Comment by hnyman** [comment]
You need a backslash for the line break, I think, so that the variable definition continues correctly

**Comment by micmac1** [comment]
> You need a backslash for the line break, I think, so that the variable definition continues correctly

Oops, fixed.

**Comment by tru7** [comment]
@micmac1 can you please check faillogs for recent builds on buildbot. Looks like it's missing location for zlib. Personally I have problems to build with ldap support but that may caused by interference with older builds of libraries in my buildroot.

**Comment by micmac1** [comment]
>@micmac1 can you please check faillogs for recent builds on buildbot.
>Looks like it's missing location for zlib. Personally I have problems
>to build with ldap support but that may caused by interference with
>older builds of libraries in my buildroot.

I've sent a pull request a few minutes ago that should address the problem locating zlib.

**Comment by micmac1** [comment]
>@micmac1 can you please check faillogs for recent builds on buildbot.
>Looks like it's missing location for zlib. Personally I have problems
>to build with ldap support but that may caused by interference with
>older builds of libraries in my buildroot.

Hello again, the packages are now built by the bots. I'd appreciate it if you could confirm that the packages built by the bots work for you, including LDAP. If there is a problem maybe you could share some details or even a simple way to reproduce. Thanks!

**Comment by tru7** [comment]
@micmac1 I quickly verified succesful building on my toolchains for ath79/mips_24kc and mpc85xx/powerpc_8540 with all modules enabled. So building looks fine with the latest commits.

I will also testrun the new packages from snapshot for mips_24kc on my device once they are available and also my own packages.

**Comment by micmac1** [comment]
On Tue, Jan 14, 2020 at 11:18:38AM -0800, Othmar Truniger wrote:
> I will also testrun the new packages from snapshot for mips_24kc on my
> device once they are available and also my own packages.

Cool, thanks a lot!

**Comment by tru7** [comment]
I successfully installed libapr, libapr and apache including all the submodules. I played successfully with normal webserver including TLS and also with mod_proxy. Looks fine for me so far.

**Comment by micmac1** [comment]
> I successfully installed libapr, libapr and apache including all the
> submodules. I played successfully with normal webserver including TLS
> and also with mod_proxy. Looks fine for me so far.

Thank you for the feedback!

**Comment by tru7** [comment]
@micmac1 woudn't it make sense to also support 19.07 with a working apache??

**Comment by micmac1** [comment]
>@micmac1 woudn't it make sense to also support 19.07 with a working
>apache??

I agree. I was thinking we wait till upstream releases .42. That would give us some time to see how it works out in snapshots. But if .42 takes more time to come out maybe we could just add a time limit of 3 or 4 weeks and then backport it. What do you think? Does that make sense?

**Comment by tru7** [comment]
sounds perfect to me.

**Comment by micmac1** [comment]
>sounds perfect to me.

Thanks, let's do it like this then.

**Comment by micmac1** [comment]
> sounds perfect to me.

Hello Othmar,

In the meantime I pushed quite a few small updates to the packages.
Could you please try out apr 1.7.0-1, apr-util-1.6.1-6 and apache
2.4.41-6? The bots have finished building packages for mips_24kc and
powerpc_8540.

I run-tested this on my router, but I sure would like a second (and
probably more thorough) test done by somebody else before I raise the
pull request to backport this to 19.07.

Kind regards,
Seb

**Comment by tru7** [comment]
I have run-tested on my tplink_archer-c7-v4 for full installation, SSL and proxy modules. Looks good to me.

**Comment by micmac1** [comment]
Super, much appreciated!
